packages/sdk/src/across/rateModel.ts

```diff
@@ -0,0 +1,158 @@
+  for (const key of expectedRateModelKeys) {
+    if (!Object.keys(rateModelFromEvent).includes(key)) {
+      throw new Error(
+        `Rate model does not contain all expected keys. Expected keys: [${expectedRateModelKeys}], actual keys: [${Object.keys(
+          rateModelFromEvent
+        )}]`
+      );
+    }
+  }
+
+  return {
+    UBar: rateModelFromEvent.UBar,
+    R0: rateModelFromEvent.R0,
+    R1: rateModelFromEvent.R1,
+    R2: rateModelFromEvent.R2,
+  };
+};
+
+/**
+ * Given an unsorted array of updated rate model events, return a dictionary mapping token addresses to sorted
+ * rate model events. This method is used internally to enforce chronological sorting of events and mapping rate models
+ * to token addresses.
+ * @param rateModelEvents Unsorted updated rate model events, each of which contains a token address, the stringified
+ * rate model for that token, and the block height of the update.
+ * @returns Dictionary mapping token addresses to chronologically sorted rate model updates.
+ */
+const createRateModelEventDictionary = (rateModelEvents: RateModelEvent[]): RateModelEventsDictionary => {
+  const updatedRateModelEventsForToken: RateModelEventsDictionary = {};
+
+  // Sort events in-place from oldest to newest:
+  rateModelEvents.sort((a, b) => {
+    if (a.blockNumber !== b.blockNumber) {
+      return a.blockNumber - b.blockNumber;
+    }
+
+    if (a.transactionIndex !== b.transactionIndex) {
+      return a.transactionIndex - b.transactionIndex;
+    }
+
+    return a.logIndex - b.logIndex;
+  });
+
+  for (const updatedRateModelEvent of rateModelEvents) {
+    // The contract enforces that all rate models are mapped to addresses, therefore we do not need to check that
+    // `l1Token` is a valid address.
+    const l1TokenNormalized = ethers.utils.getAddress(updatedRateModelEvent.l1Token);
+    if (!updatedRateModelEventsForToken[l1TokenNormalized]) updatedRateModelEventsForToken[l1TokenNormalized] = [];
+
+    // We assume that events are returned from oldest to newest, so we can simply push events into the array and
+    // and maintain their time order.
+    updatedRateModelEventsForToken[l1TokenNormalized].push({
+      blockNumber: updatedRateModelEvent.blockNumber,
+      rateModel: updatedRateModelEvent.rateModel,
+    });
+  }
+
+  return updatedRateModelEventsForToken;
+};
+
+/**
+ * Return the rate model for L1 token set at the block height.
+ * @param rateModelEvents Reformatted unordered UpdatedRateModel events.
+ * @param l1Token L1 token address to get rate model for.
+ * @param blockNumber Block height to get rate model for.
+ * @returns Rate model object.
+ */
+export const getRateModelForBlockNumber = (
+  rateModelEvents: RateModelEvent[],
+  l1Token: string,
+  blockNumber?: number
+): RateModel => {
+  const l1TokenNormalized = ethers.utils.getAddress(l1Token);
+  const rateModelDictionary = createRateModelEventDictionary(rateModelEvents);
```

Comment:
Theoretically, how many times could this rate model dictionary be created per run? Does it get rebuilt every time we need to grab a rate model for a relay? If there were 100 relays within the last 2 hours, would we create this 100 times?

Thoughts on making this a persistent class where the rate model events are passed on construction?

Reply:
Yeah it does get rebuilt every time a new rate model is constructed, I can change to a persistant class with storage

Reply:
> Does it get rebuilt every time we need to grab a rate model for a relay?

Yes, good point.

I've added a persistent `RateModelDictionary` class in this PR, I think this means that 1 dictionary would be created per serverless run. [PR](https://github.com/UMAprotocol/protocol/pull/3673/commits/fa6270e377eb61e28c61407ae31ed03c6a2647b3)